### ISSUE TITLE
int(INF) evaluates to negative number

### ISSUE THREAD

**Issue by simenhs** [godotengine/godot#100551]
### Tested versions

Reproducable in : v4.3.stable

### System information

Windows 11 

### Issue description

Casting the float INF to int with int(INF) returns -9223372036854775808 and not 9223372036854775807 witch is the largest int. int(-INF) also returns -9223372036854775808 which make sense to me. 

### Steps to reproduce

Print(int(INF)) 

### Minimal reproduction project (MRP)

[int(inf)bug.zip](https://github.com/user-attachments/files/18177991/int.inf.bug.zip)

**Comment by hpvb** [comment]
Casting a NaN to an int maybe shouldn't be possible at all. This is both mathematically and logically nonsense.

But I don't think this is a "bug" as such, neither really makes sense.

**Comment by Mickeon** [comment]
> Casting a NaN to an int maybe shouldn't be possible at all. This is both mathematically and logically nonsense.

There's no mention of NaN here, this is specifically about INF.
However, INF, -INF and NaN all return the minimum integer value in GDScript. This makes me think that this behavior is **not intentional**, because most functions in Godot return the type's default value on failure (in this case it should be `0`).

Now, what should happen can be discussed. Still, the lack of an error or a warning should be concerning.

**Comment by Mickeon** [comment]
In C and C++, casting INF, -INF, and NaN to an integer is an entirely invalid operation.

However, Java does something unique. Casting these floats to an integer returns:
- the maximum integer value for INF
- the minimum integer value for -INF.
- `0` for NaN.

Given that GDScript happens to be friendly in this regard, the equivalence between float and integer Vector types (`Vector2.INF`, `Vector2.MAX` and `Vector2i.MIN` constants), and it seems reasonable to assume the ideal behaviour may be similar to Java's. For most games the biggest integer number may as well be infinity.

I repeat, if we do not want to argue about this, for the reasons stated above it should **at least** return `0` in all of the situations above. At most there should be a warning about this, because in most languages this would fail the entire application.

**Comment by Ivorforce** [comment]
I'm very surprised, apparently this is a C++ thing:

https://onlinegdb.com/i9PSoC2Tv

The relevant Godot code simply hands off to C-style conversion: 

https://github.com/godotengine/godot/blob/6395450b10d73bc3515763c7bbd1b2f5b7425d10/core/variant/variant.cpp#L1498-L1499

**Comment by hpvb** [comment]
In my opinion any behavior is a bug in one way or another. If we get people trying to know if something was an inf or not by looking at the integer value we've already "lost", we have is_inf and such that really should be used.

**Comment by hpvb** [comment]
> Given that GDScript happens to be friendly in this regard, the equivalence between float and integer Vector types (`Vector2.INF`, `Vector2.MAX` and `Vector2i.MIN` constants), it seems reasonable to assume the ideal behaviour may be similar to Java's. For most games the biggest integer number may as well be infinity.
> 
> I repeat, if we do not want to argue about this, for the reasons stated above it should **at least** return `0` in all of the situations above. At most there should be a warning about this, because in most languages this would fail the entire application.

I agree from a user-friendliness point of view, however it would introduce a new check in a codepath which could be relatively hot for certain games. I think what I would like to see is a new method then that does things in this way, so the cost is opt-in for game developers.

Would this be a reasonable compromise?

**Comment by Mickeon** [comment]
I feel like usage of a separate method is too niche. There's no real demand for it, either.
In my uttermost ignorance of the topic the warning could only be sent in debug builds.
The behavior with `NaN`, `INF` and `-INF` needs to be documented for sure, regardless of what it is.

Yeah, I got nothing else to say.

**Comment by greenfox1505** [comment]
I ran into this bug today. I was running a loop looking for a minimum number. I created a `var best:int = INF` expecting to get an int max. If I had just use `var best = INF`, this wouldn't be an issue. Could not figure out what was wrong until I wrote this test case. Instead I got `INT64_MIN`. It's for a tool script (`EditorImportPlugin`) so, running a debugger was not easy to do. I didn't figure out the issue until I wrote this.

```
var inf:int = INF
var max:int = INT64_MAX
print("inf = ",inf)
print("max = ",max)
```

> In my opinion any behavior is a bug in one way or another. 

Yeah, maybe. Casting INF doesn't have a lot of good options. But casting it to `INT64_MIN` is probably the **most wrong** option.

**Comment by Arecher** [comment]
The Java implementation is definitely most intuitive for users that simply use `INF` and `-INF` as **really big number** and **really small number**. While limiting `INF` to the largest maximum int that is supported is a logical enough limitation (and therefore easier to debug when you run into an issue), receiving the smallest int possible when inputting `INF` is so unexpected it can only lead to unnecessary confusion.

A warning would be good regardless.

**Comment by Arecher** [comment]
Duplicate of https://github.com/godotengine/godot/issues/33017 (and a few others). Although the discussion here has gone the furthest.

**Comment by AThousandShips** [comment]
I'd say no integer value being given as a conversion from infinity would be correct and adding any overhead to attempt to "correct" this would be wasteful, I think a note in the documentation is sufficient

> The Java implementation is definitely most intuitive for users that simply use INF and -INF as really big number and really small number.

I'm not sure we should make decision on implementation based on incorrect application of concepts

**Comment by Arecher** [comment]
```
var lowest_value = INF
var my_values = []
for i in my_values:
    if i < lowest_value:
        lowest_value = i
```

Seems like a common use-case of INF to me, and essentially boils down to **really big number**. In which case I'd argue turning it into an int which is still the **largest possible** number would be more correct than turning it into the **smallest possible** number.

**Comment by AThousandShips** [comment]
But that's wrong, that's not what infinity *is*, people might use it like that but it's a fundamental misunderstanding of what `INF` is

Additionally converting it to `INT64_MAX` would make it behave incorrectly as `INT64_MAX` would not be *less than* `int(INF)`

But to be clear: infinity is not the same as "a really big value", it isn't a number or a value, it's a *concept* in calculus. Saying that any number is equal to infinity is incorrect conceptually (things can *tend* to infinity in a limit for example though), this is why `INF - INF` results in `NaN`, as well as `INF * 0.0`

And additionally, the fact that using `float` `INF` in that way doesn't mean it makes sense to use the result of casting `INF` to `int` in that way, and relying on the value of casting `INF` to `int` being any specific value is a bad idea

It being confusing that it doesn't have a defined value when cast to an `int` might be reasonable (though it isn't really incorrect as there isn't a correct value, so it *should* be undefined), it doesn't really make sense to expect it to have a well defined value and there aren't really any valid use cases of that cast I'd say

**Comment by zhengxiaoyao0716** [comment]
I encountered a strange type‑conversion behavior related to Array[float] casting and Vector2 construction.
I am not sure where the actual problem is, but the behavior is consistently reproducible.

Test code:
```gdscript
  var arr := [-1]
  var num := (arr as Array[float])[0]
  assert(not is_nan(num))
  prints("num:", num, num as float, num as int)
  prints("typeof(num):", type_string(typeof(num)))
  var vec := Vector2(num, 0.0)
  prints("vec:", vec)
  prints("is_nan(vec.x):", is_nan(vec.x))
  prints("vec.x", vec.x, vec.x as float, vec.x as int)

```

Actual output:
```
num: -1 -1.0 -1
typeof(num): int
vec: (nan, 0.0)
is_nan(vec.x): true
vec.x nan nan -9223372036854775808

```

**Comment by blanchardj-dev** [comment]
The behaviour shown is most probably the effect of undefined floating point operations being left to the hardware implementation. On intel and AMD I think non-finite numbers are always cast to the largest negative value meanwhile on ARM the behaviour can vary from chip to chip depending on the FPU implementation.

Also of importance, when a cast can be optimized out, it's possible to see different behaviour between optimization levels when the compiler can precalculate the result, the value returned then depends on the compiler implementation instead of the hardware.

The only ways to guard against that kind of behaviour that I know of, is either to explicitly validate the value with isfinite() or checking if an error is raised after using lrint() or similar.

This is also relevant to the discussion of #101417

(Sorry to comment out of the blue I was just browsing random issues and could not help myself but add my 2c)
